src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFilter.cs

```diff
@@ -0,0 +1,114 @@
+        private const string Alias = "Microsoft.Targeting";
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="loggerFactory">A logger factory for creating loggers.</param>
+        public ContextualTargetingFilter(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory?.CreateLogger<ContextualTargetingFilter>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Performs a targeting evaluation using the provided <see cref="TargetingContext"/> to determine if a feature should be enabled.
+        /// </summary>
+        /// <param name="context">The feature evaluation context.</param>
+        /// <param name="targetingContext">The targeting context to use during targeting evaluation.</param>
+        /// <returns>True if the feature is enabled, false otherwise.</returns>
+        public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context, ITargetingContext targetingContext)
+        {
+            TargetingFilterSettings settings = context.Parameters.Get<TargetingFilterSettings>() ?? new TargetingFilterSettings();
+
+            //
+            // No audience targeted, the feature will be off
+            if (settings.Audience == null)
+            {
+                _logger.LogWarning($"The '{Alias} feature filter does not have an audience for the feature '{context.FeatureName}'.");
+
+                return Task.FromResult(false);
+            }
+
+            //
+            // Check if the user is being targeted directly
+            if (targetingContext.UserId != null &&
+                settings.Audience.Users != null &&
+                settings.Audience.Users.Any(user => targetingContext.UserId.Equals(user)))
+            {
+                return Task.FromResult(true);
+            }
+
+            //
+            // Check if the user is in a group that is being targeted
+            if (targetingContext.Groups != null &&
+                settings.Audience.Groups != null)
+            {
+                foreach (string group in targetingContext.Groups)
+                {
+                    GroupRollout groupRollout = settings.Audience.Groups.FirstOrDefault(g => g.Name.Equals(group));
+
+                    if (groupRollout != null)
+                    {
+                        string audienceContextId = $"{targetingContext.UserId}\n{context.FeatureName}\n{group}";
+
+                        if (IsTargeted(audienceContextId, groupRollout.RolloutPercentage))
+                        {
+                            return Task.FromResult(true);
+                        }
+                    }
+                }
+            }
+
+            //
+            // Check if the user is being targeted by a default rollout percentage
+            string defaultContextId = $"{targetingContext.UserId}\n{context.FeatureName}";
```

Comment:
there might be the case with no userid but group info, we might use both or one of them which exist

Reply:
UserId being null is the same as userId of empty string in this case. Groups don't play a factor in the default rollout percentage evaluation. It is based purely on user.